### ISSUE TITLE
Throttle filter is causing Logstash to crash under heavy load

### ISSUE THREAD

**Issue by AGirin** [logstash-plugins/logstash-filter-throttle#12]
Hi! Opening new issue as requested:
- Version: 2.3.4
- Operating System: CentOS 6.6

We observed that if we have thousands throttle events Logstash crashes with this error:

`{:timestamp=>"2016-07-28T11:34:34.318000-0500", :message=>"Exception in pipelineworker, the pipeline stopped processing new events, please check your filter configuration and restart Logstash.", "exception"=>#<NoMethodError: undefined method`latest=' for nil:NilClass>`

We are using @frapex code (I believe it was finally merged) and just today had an issue in production where 3 out of 4 logstash servers crashed and caused issues in our pipeline. 

Unfortunately logstash creates HUGE log line when it crashes and I am not sure if it is a good idea to paste it here but the end of the lines before the crash look like this:

"current_call"=>"[...]/vendor/bundle/jruby/1.9/gems/logstash-core-2.3.4-java/lib/logstash/pipeline.rb:309:in `synchronize'"}]}>, :level=>:warn}

Basically we send all alerts created by logstash to pagerduty (https://github.com/logstash-plugins/logstash-output-pagerduty). My guess something happens to pipeline when we have to send and throttle a lot of alerts. Please let me know if you need more info and I will try to provide it.

**Comment by frapex** [comment]
I identified the issue you are having under very heavy load. It happens when a timeslot cache instance is deleted during flush, while at the same time an event is being processed that tries to get the timeslot cache that was just deleted.

Code that deletes the timeslot cache:

```
        @key_cache.delete(key)               # delete expired timeslot cache
```

Exception triggered at:

```
    timeslot_cache = @key_cache[key]         # get timeslot cache
    timeslot_cache.latest = epoch            # update to latest epoch
```

I'll look into this.

**Comment by frapex** [comment]
@AGirin Can you please test the patch below and let me know how it works out for you?

patch:

```
--- a/lib/logstash/filters/throttle.rb
+++ b/lib/logstash/filters/throttle.rb
@@ -226,11 +226,14 @@ class LogStash::Filters::Throttle < LogStash::Filters::Base
     period = 60 if period == 0               # fallback if unparsable
     epoch = event.timestamp.to_i             # event epoch time

-    @key_cache.compute_if_absent(key) do     # initialise timeslot cache
-      ThreadSafe::TimeslotCache.new(epoch)   # and add to key cache
+    while true
+      @key_cache.compute_if_absent(key) do   # initialise timeslot cache
+        ThreadSafe::TimeslotCache.new(epoch) # and add to key cache
+      end
+      timeslot_cache = @key_cache[key]       # get timeslot cache
+      break unless timeslot_cache.nil?       # break loop on success
     end

-    timeslot_cache = @key_cache[key]         # get timeslot cache
     timeslot_cache.latest = epoch            # update to latest epoch

     # find target timeslot
```

And you should definitely tune the plugin config: decrease max_age and/or increase max_counters.

**Comment by AGirin** [comment]
@frapex - will test and let you know. Thanks a lot for the quick response!

**Comment by AGirin** [comment]
So far no issues. Tested in Staging and Production - no problems yet. Will continue and let you know if everything looks good.

**Comment by frapex** [comment]
Great, keep it running for another week or so.

**Comment by frapex** [comment]
@AGirin pull request created

**Comment by frapex** [comment]
@suyograo this issue can now be closed, thanks for merging the fix.
